Improve date format consistency

davidrapson opened the pull request:
This PR refactors how we display dates to be consistent in how ranges are handled. Helps with multi-day events, and improves consistency in general.

![screen shot 2015-06-30 at 15 48 01](https://cloud.githubusercontent.com/assets/123386/8433755/a4582aac-1f3f-11e5-8ff4-2bb3170053e4.png)
![screen shot 2015-06-30 at 15 47 49](https://cloud.githubusercontent.com/assets/123386/8433756/a4597b6e-1f3f-11e5-9534-a0a165d82fcf.png)
![screen shot 2015-06-30 at 15 48 09](https://cloud.githubusercontent.com/assets/123386/8436152/0d52059c-1f4e-11e5-84c4-f5ecaa6ce955.png)

@rtyley

davidrapson commented on frontend/app/views/support/Dates.scala:
```diff
@@ -59,18 +59,31 @@ object Dates {
 
-  def dateRange(start: DateTime, end: DateTime): Range = {
-    def dateToks(dt: DateTime) = dt.toString("EEEE d MMMMM YYYY").split(" ").reverse
+  private def buildDateRange(start: DateTime, end: DateTime, fmt: String): Range = {
```
@rtyley I pulled this out just because the code would've been repeated across `dateRange` and the new `shortDateRangeString`. Not quite sure what the code is doing to give it a meaningful name. Any thoughts on how to better refactor this appreciated.

rtyley replied:
The code is pretty scary to look at, not surprised you thought it weird. The tests give some indication of what's going on: https://github.com/guardian/membership-frontend/blob/ebf425fab2/frontend/test/views/support/DatesTest.scala#L7-L42

The code is trying to display a date range (start-end) in a human-friendly way that is pleasantly concise by not unnecessarily repeating any parts of the string. Eg. do not do: 

```
Thursday 6 November 2014, 10.20am - Thursday 6 November 2014, 12.30pm
Thursday 6 November 2014 - Saturday 8 November 2014
```

...when you can say:

```
Thursday 6 November 2014, 10.20am - 12.30pm
Thursday 6 - Saturday 8 November 2014
```

If the start & end are on the same day, we display the date with the start-end times, otherwise we omit the time and use the second form, which removes duplication by truncating the right-hand-side of the 'start' string by all the parts that are already in the 'end' string.

I hope that explains it a little, though I'm not sure if it immediately suggests a better name.

davidrapson replied:
The eventual behaviour was clear (and very nice), just not what the particular part of code controlled so that helps a bit.

davidrapson commented:
@rtyley Added a few tests around the new method, same scenarios as dateRange. Look OK to you?

davidrapson commented:
@rtyley Cleaned up the naming, switched to use `Interval` and moved the `interval.start.withTimeAtStartOfDay() == interval.end.withTimeAtStartOfDay()` line to and `implicit class RichInterval` which just worked which is magic :star2:

rtyley commented on frontend/test/views/support/DatesTest.scala:
```diff
@@ -1,43 +1,62 @@
+//      val interval = new Interval(new DateTime(2014, 11, 6, 15, 0),  new DateTime(2014, 12, 8, 13, 0))
+//      Dates.dateTimeRange(interval) mustEqual "Thursday 6 November–Monday 8 December 2014"
+//    }
+//
```
Remove commented out code?

davidrapson replied:
@rtyley Already a head of you. I did not mean to leave these commented out.

rtyley commented:
> @rtyley Cleaned up the naming, switched to use Interval and moved the interval.start.withTimeAtStartOfDay() == interval.end.withTimeAtStartOfDay() line to and implicit class RichInterval which just worked which is magic

Maaaaaagic!

![](https://media4.giphy.com/media/ejK2NEJQFG9fW/200.gif)

:+1:

davidrapson commented:
![themagicwaswithinyouallalong](https://cloud.githubusercontent.com/assets/123386/8457100/569e468e-2007-11e5-8096-1095ee766e17.gif)
